### ISSUE TITLE
Please add more details on "Token is unsafe!" message

### ISSUE THREAD

**Issue by ChristianStadelmann** [freeotp/freeotp-android#287]
When scanning a QR code from Microsoft's mfasetup on https://aka.ms/mfasetup, I get this warning:
> Token is unsafe!
> 
> The token you are attempting to add contains weak cryptographic parameters. Use of this token is stronly discouraged! Please alert your token provider.
> 
> [Cancel] [Add Anyway]

It might be useful to have more details on why the token is unsafe, e.g. some text about the algorithm (cipher, key exchange mechanism, parameters, …). If possible, it would also be nice to tell why the token is unsafe.

Version info:
FreeOTP 2.0 (24) from F-Droid repository on Android 11.

**Comment by yanivhs** [comment]
I'm using a sha512 tokens, and also got this message...

Looking at the app code, you can see it expect it to use uppercase letters while in the examples they are in lowercase.

**Comment by ChristianStadelmann** [comment]
> Looking at the app code, you can see it expect it to use uppercase letters while in the examples they are in lowercase.

Thanks for this hint, I've created #288 for that.

**Comment by ninernet** [comment]
I recently (3 January) factory-reset my phone and reinstalled the app. I didn't even look for options to back up what I already had set up in the app (and I don't think they existed in the version I had installed at the time) as I assumed that getting set up again with the vendors that require the use of an authenticator app would be straightforward. How wrong I was!

I only had two vendors set up in the app. The first one I tried to re-set up is a company where I expect issues on an almost daily basis, and I got the "Token is unsafe!" warning. I brought it to their attention last week and I am still awaiting a response on how I should proceed. With my bank today I did *not* expect this warning, and after half an hour of trying to help me they gave up and opened a ticket with their IT department.

So yes, per the original poster some additional information would probably be useful. But since this is affecting a bank and Microsoft (as opposed to mom-and-pop outfits), I'm actually wondering if this might be a bug. Supposedly I will hear from my bank tomorrow, but in advance I've been trying to find information on this issue so that I can look semi-intelligent when they call.

**Comment by mokraemer** [comment]
It looks like freeOTP requires that tokens have at least 128bit (26 base32 coded digis). Otherwise it is considered unsecure.
It would be very helpful to show this hint. 80 bits (16 base32 digits) were accepted before and are still by the ios app.

**Comment by justin-stephenson** [comment]
> It looks like freeOTP requires that tokens have at least 128bit (26 base32 coded digis). Otherwise it is considered unsecure. It would be very helpful to show this hint. 80 bits (16 base32 digits) were accepted before and are still by the ios app.

Yes, due to https://www.ietf.org/rfc/rfc4226.txt algorithm compliance requirements but I agree we should more clearly state this in the error message. We should also establish uniformity with FreeOTP iOS, and perhaps provide an option to ignore this insecure warning and add the token anyway.

**Comment by mokraemer** [comment]
it is ok, to force better keys. But e.g. [php gangsta](https://github.com/PHPGangsta/GoogleAuthenticator) has a default length of 80 bit. And I was confused that now I was warned about unsafe "algorithm" and did not know "only" the length should be extended >128

**Comment by kingma-sbw** [comment]
I'm not sure if the hint to swap the OTP app just for GH is an acceptabel solution.

**Comment by telephon** [comment]
Also, the QR-code displayed after the scan doesn't visually match the original QR-code – this is very confusing. I would expect such a behavior from an app that has been tinkered with.

**Comment by akostadinov** [comment]
Yeah, would be good to clarify this message and see if it should apply to TOTP at all.

**Comment by m8ram** [comment]
Now that Github is requiring 2FA I also ran into this. I opened the following issue with github: https://github.com/orgs/community/discussions/172923 .
It would indeed help if the message explained exactly what parameters are considered weak to provide more feedback to the providers.
Having public documentation on what FreeOTP considers secure or not would be helpful as well.
This issue is the only place I found so far that mentions the minimum of 128 bits.

**Comment by KUGA2** [comment]
Just ran into this with my companies MS account. Of course our company IT blames this app.

**Comment by jarovo** [comment]
Hello. According to [theinckdude](https://www.reddit.com/r/yubikey/comments/rb3u3d/comment/hnmqrwm/) the usage of SHA-1 in TOTP application is safe. Quoting:

 >        you can make an arbitrary input match a given SHA1 hash
> 
>  This is a property called preimage resistance, and no, you can't. You can generate two inputs that hash to 
>  the same SHA1 hash, called a collision attack, as long as you get to control both of those two inputs yourself.
>  This is dangerous in situations like document signing, where you can create two different documents that
> hash to the same value, so a signature computed on the "good" document is also valid for the "evil" document.
>  But it has no impact on TOTP.
> 
[SHA-1 on medium](https://medium.com/@alecmuffett/no-the-shattered-sha-1-attack-does-significantly-impact-tor-yet-f64f859cc287)

So how big is the deal of using obsolete SHA-1 in TOTP application -- should the warning be even displayed?
